_posts/2019-05-23-quarkus-and-angular-ui-development-mode.adoc

```diff
@@ -0,0 +1,391 @@
+
+[source,java]
+----
+@NgModule({
+  declarations: [
+    AppComponent,
+    DefaultComponent,
+    OtherComponent,
+    RestComponent,
+    ClientCallbackComponent
+  ],
+  imports: [
+    BrowserModule,
+    AppRoutingModule,
+    HttpClientModule
+  ],
+  providers: [],
+  bootstrap: [AppComponent]
+})
+export class AppModule { }
+
+----
+
+
+All this does is that when we go to the root of the application, we will go to DefaultComponent. `DefaultComponent` has routerLinks to `OtherComponent` and `RestComponent` (for routerLinks Angular does not hit the back-end), and a normal link to `/servlet/make-external-call`.
+
+If we go to `/other`, we end up in `OtherComponent` which just has a link back to `DefaultComponent`.
+
+If we go to `/rest`, we end up in `RestComponent` which displays data from the `SampleResource` we saw above and also has a link back to `DefaultComponent`.
+
+Finally if we go to `/clientCallback` (which is triggered via `/servlet/callback`  in `/SampleServlet` we end up in `ClientCallbackComponent`.
+
+====== Illustrating the problem
+Now package and start the application by running:
+[source,bash]
+----
+$mvn package quarkus:dev -Dui.dev
+----
+If you go to http://localhost:8080 you will get a page with links to `Other`, `Rest` and `Default`. Click on the `Other` and `Rest` ones, and it should all work.
+
+However while in the `Other` component, so that the address in the browser is http://localhost:8080/other, if you try to refresh the page you will end up with the following error message:
+
+`RESTEASY003210: Could not find resource for full path: http://localhost:8080/other`
+
+Also, if we go back to http://localhost:8080, and click the `External` link, we will see a similar message.
+
+We will fix these in the next step. Stop Quarkus before going to the next step.
+
+==== Step 4 - Servlet filter to forward UI paths to Angular
+The diff for this step can be found https://github.com/kabir/blog-quarkus-ui-development/compare/step-3..step-4[here].
+
+We saw in the previous step that when trying to go directly to a route within the Angular application we end up hitting the server which cannot find a matching REST endpoint, which is not what we would expect.
+
+To deal with this I add a servlet filter:
+
+[source,java]
+----
+@WebFilter(urlPatterns = "/*")
+public class IndexHtmlFilter extends HttpFilter {
+
+    private static final Pattern FILE_NAME_PATTERN = Pattern.compile(".*[.][a-zA-Z\\d]+");
+
+    public void doFilter(ServletRequest req, ServletResponse res, FilterChain chain) throws IOException, ServletException {
+        HttpServletRequest request = (HttpServletRequest) req;
+        HttpServletResponse response = (HttpServletResponse) res;
+
+        if (isIndexHtmlRequest(request)) {
+            request.getRequestDispatcher("/").forward(request, response);
+        } else {
+            chain.doFilter(request, response);
+        }
+    }
+
+    private boolean isIndexHtmlRequest(HttpServletRequest request) {
```

Comment:
I would change the method name, it's confusing.

`assumeAngularRoute()`  or `isAngularRoute` sound better to me